addons/l10n_my_edi_extended/models/account_move.py

```diff
@@ -0,0 +1,188 @@
+                    'raw': xml_content,
+                    'mimetype': 'application/xml',
+                    'res_model': move._name,
+                    'res_id': move.id,
+                    'res_field': 'l10n_my_edi_file',  # Binary field
+                })
+                move.invalidate_recordset(fnames=['l10n_my_edi_file_id', 'l10n_my_edi_file'])
+
+    def _l10n_my_edi_prepare_moves_to_send(self):
+        AccountMoveSend = self.env['account.move.send']
+        xml_contents = defaultdict(list)
+        moves = self.env['account.move']
+        for move in self:
+            if not move.l10n_my_invoice_need_edi or move.l10n_my_edi_state:
+                continue
+
+            moves |= move
+
+            if move.l10n_my_edi_file:
+                xml_content = base64.b64decode(move.l10n_my_edi_file).decode('utf-8')
+            else:
+                xml_content, errors = move._l10n_my_edi_generate_invoice_xml()
+                if errors:
+                    raise UserError(AccountMoveSend._format_error_text({
+                        'error_title': _('Error when generating MyInvois file:'),
+                        'errors': errors,
+                    }))
+                xml_content = xml_content.decode('utf-8')
+            xml_contents[move] = xml_content
+        return moves, xml_contents
+
+    def _l10n_my_edi_send_to_myinvois(self, moves, xml_contents):
+        AccountMoveSend = self.env['account.move.send']
+        if moves and xml_contents:
+            errors = moves._l10n_my_edi_submit_documents(xml_contents)
+
+            if errors:
+                for move in moves:
+                    move.with_context(no_new_invoice=True).message_post(body=AccountMoveSend._format_error_html({
+                        'error_title': _('Error when sending the invoices to the E-invoicing service.'),
+                        'errors': errors[move],
+                    }))
+
+            # At this point we will need to commit as we reached the api, and we could have a mix of failed and valid invoice.
+            if moves._can_commit():
+                self._cr.commit()
+
+            # We already logged the details on the invoice(s) and saved the api results. If we send a single invoice, we can safely raise now.
+            if errors and len(moves) == 1:
+                raise UserError(AccountMoveSend._format_error_text({
+                    'error_title': _('Error when sending the invoices to the E-invoicing service.'),
+                    'errors': errors[moves],
+                }))
+
+    def _l10n_my_edi_get_status(self, moves):
+        AccountMoveSend = self.env['account.move.send']
+        retry = 0
+        errors, any_in_progress = moves._l10n_my_edi_fetch_updated_statuses()
+        while any_in_progress and retry < 2:
+            time.sleep(1)  # We wait a second before retrying.
+            errors, any_in_progress = moves._l10n_my_edi_fetch_updated_statuses()
```

Comment:
Won't this forget about potential errors if found previously?

Reply:
In practice, it's not very important because we do this for a single submission, and one submission = one API call to get the status.
So if we "raise" errors the first time, I believe the same error would still be there the third time.